File: src/main/scala/drivers/MongoDriver.scala
package drivers

import entities.Entities.Specification
import org.bson.codecs.configuration.CodecRegistries.{fromProviders, fromRegistries}
import org.bson.codecs.configuration.{CodecProvider, CodecRegistry}
import org.mongodb.scala.MongoClient.DEFAULT_CODEC_REGISTRY
import org.mongodb.scala.bson.codecs.Macros.createCodecProvider
import org.mongodb.scala.{MongoClient, MongoCollection, MongoDatabase}

import scala.concurrent.Future

object MongoDriver {
    import VertxDriver.executionContext

    val mongoClient: MongoClient = MongoClient()
    val db: MongoDatabase = mongoClient.getDatabase("mas-web-ide").withCodecRegistry(Codec.codecRegistry)
    implicit val specifications: MongoCollection[Specification] = db.getCollection("specifications")

    implicit val codecProvider: CodecProvider = createCodecProvider[Specification]()

    object Codec {
        val codecRegistry: CodecRegistry = fromRegistries(fromProviders(
            classOf[Specification],
        ), DEFAULT_CODEC_REGISTRY)
    }

    def getSpecifications(specifications: MongoCollection[Specification] = MongoDriver.specifications): Future[List[Specification]] =
        specifications.find().toFuture().map(_.toList)
}

File: src/main/kotlin/drivers/MongoDriver.kt
package drivers

import com.mongodb.ConnectionString
import entities.Entities.MoiseSpecification
import entities.Entities.Specification
import kotlinx.coroutines.runBlocking
import org.bson.types.ObjectId
import org.litote.kmongo.coroutine.coroutine
import org.litote.kmongo.eq
import org.litote.kmongo.id.toId
import org.litote.kmongo.reactivestreams.KMongo

class MongoDriver {
    companion object {
        val CONNECTION_STRING: String = System.getenv("MONGO_CONNECTION_STRING") ?: "mongodb://localhost:27017"
    }

    private val client = KMongo.createClient(ConnectionString(CONNECTION_STRING)).coroutine
    private val database = client.getDatabase("mas-organizations")
    private val specifications = database.getCollection<Specification>("specifications")
    private val moiseSpecs = database.getCollection<MoiseSpecification>("moiseSpecifications")

    fun getSpecifications(): List<Specification> = runBlocking {
        specifications.find().toList()
    }

    fun addSpecification(s: Specification): Boolean = runBlocking {
        specifications.save(s)?.wasAcknowledged() ?: false
    }

    fun updateSpecification(id: String, s: Specification): Boolean = runBlocking {
        specifications.updateOneById(ObjectId(id).toId<Specification>(), s).wasAcknowledged()
    }

    fun deleteSpecification(id: String): Boolean = runBlocking {
        specifications.deleteOneById(ObjectId(id).toId<Specification>()).wasAcknowledged()
    }

    fun addMoiseSpec(s: MoiseSpecification): Boolean = runBlocking {
        moiseSpecs.save(s)?.wasAcknowledged() ?: false
    }

    fun getMoiseSpec(name: String): MoiseSpecification? = runBlocking {
        moiseSpecs.findOne(MoiseSpecification::name eq name)
    }

    fun updateMoiseSpec(s: MoiseSpecification) = runBlocking {
        moiseSpecs.updateOne(MoiseSpecification::name eq s.name, s)
    }
}
